File: scripts/python/clean_old.py
#   Noise reduction using spectral noise gating.
#   Reads a CSV file containing results of noise analysis, and proceeds
#   to reduce the noise of a song.
#   
#   Version 1.0: Noise reduced, and time-smoothing and frequency smoothing applied.
#
#	args:
#		sys.argv[1]: 	Dirty file path
#		sys.argv[2]: 	Noise year
#		sys.argv[3]: 	Noise profile
#		sys.argv[4]: 	Reduce gain
#		sys.argv[5]: 	Smoothing bands
#

#	=====================================================================
#	Requirements
#	=====================================================================

import sys
import time
import math
import magic
import os
import numpy as np
import scipy.fftpack as fftpack
import scipy.io.wavfile as wav
from pydub import AudioSegment
mime = magic.Magic(mime = True)

#	=====================================================================
#	Some validation
#	=====================================================================

#	If not enough args, return
if len(sys.argv) < 6 :
	msg = 	'Noise reduction using spectral noise gating.\nReads a CSV file containing results of noise analysis, and proceeds\nto reduce the noise of a song.\n\nVersion 1.0: Noise reduced, and time-smoothing and frequency smoothing applied.\n\nArguments (all of them are compulsory):\n\tsys.argv[1]: 	Dirty file path\n\tsys.argv[2]: 	Noise year\n\tsys.argv[3]: 	Noise profile\n\tsys.argv[4]: 	Reduce gain\n\tsys.argv[5]: 	Smoothing bands'
	sys.exit(msg)

#	Args
input_original_file 	= sys.argv[1]
noise_year 				= sys.argv[2]
noise_profile 			= sys.argv[3]
reduce_gain 			= sys.argv[4]
smoothing_bands 		= sys.argv[5]

#	Debugging logs
print('Args: %s' % ', '.join(sys.argv))

#	If the file doesn't exist
if not os.path.isfile(input_original_file) :
	sys.exit('File doesn\'t exist')

#	=====================================================================
#	Conversion to WAV
#	=====================================================================

#	Input (generally non-WAV)
input_original_name, input_original_extension = os.path.splitext(input_original_file)
input_original_mime = mime.from_file(input_original_file)
input_original_format = input_original_extension.replace('.', '', 1)

#	Converted input (WAV)
input_converted_format = 'wav'
input_converted_name = os.path.basename(input_original_name)
input_converted_file = input_original_file.replace(input_original_extension, '.wav')

#	Original output (WAV)
output_original_file = input_converted_file.replace('dirty', 'clean', 1)

#	Converted output
output_converted_file = output_original_file.replace('.wav', input_original_extension, 1)
output_converted_format = input_original_format

#	Noise info
#	noise_path = os.path.abspath(os.path.join(os.getcwd(), '../../', 'files/noise/profiles', noise_year, noise_profile + '.csv'))
noise_path = os.path.abspath(os.path.join('files/noise/profiles', noise_year, noise_profile + '.csv'))

#	WAV MIME types
wav_extensions = ['.wav']

#	Time measure
start_time = time.time()

#	If the input file is not yet a WAV
if not input_original_extension in wav_extensions :
	#	CONvert to WAV
	AudioSegment.from_file(input_original_file, input_original_format).export(input_converted_file, format = input_converted_format)
	print('Conversion to WAV took %.4f seconds' % (time.time() - start_time))
	#	Time measure
	start_time = time.time()

#	=====================================================================
#	Noise reduction
#	=====================================================================

#	Read uploaded song
Fs, Song = wav.read(input_converted_file)
Song = Song.astype(float)
song_norm_factor = np.amax((math.fabs(np.amax(Song)), math.fabs(np.amin(Song))))
Song = Song / song_norm_factor

#	Read noise statistics
NoisePowers = np.genfromtxt(noise_path, delimiter = ',')
NoisePowers = 0.16 * NoisePowers

#   Matrix dimensions
songchannels = Song.ndim
songlength = len(Song)
noisechannels = NoisePowers.ndim
FFTsize = len(NoisePowers)

#	Parameters depending on the previous ones
W = FFTsize
MSS = W / 2
ReduceLevelUN = 1.0
ReduceLevelUN = 1 / (10 ** (float(reduce_gain) / 10))
times = 0

#	Window
Window = np.hanning(W)

#	Initialize
iterations = math.ceil(songlength / MSS)
NewSong = np.zeros((songlength + W, songchannels))

#	Print status
print('Step 1: Taking statistics from %d samples...' % songlength)

#	Analyze song
for j in range(0, songchannels) :

	#	Print channel and initialize progress
	print('\tChannel no. %d' % (j + 1))
	# total = math.floor(songlength / MSS)
	# progress = 0
	count = 0
	
	for i in range(0, songlength, MSS) :
		
		#	Calculate end of song and noise
		songend = np.amin((i + W, songlength - 1))
		
		#	Sample
		SongSample = Song[i : songend, j]

		#	If we reached the end
		if len(SongSample) < W :
			#	Fill with zeros until size equals W
			Zeros = np.zeros(W, dtype = int)
			Zeros[0 : len(SongSample)] = SongSample
			SongSample = Zeros

		#	Windowed sample
		WindowedSample = SongSample * Window
		
		#	Compute FFT
		SampleTransform = fftpack.fft(WindowedSample)
		#	Calculate power
		Power = abs(SampleTransform) ** 2

		#	Apply gains to values under threshold
		for k in range(0, len(Power)) :
			if Power[k] <= NoisePowers[k, j] :
				SampleTransform[k] = SampleTransform[k] * ReduceLevelUN
		
		#	Inverse FFT
		ProcessedSample = np.real(fftpack.ifft(SampleTransform))

		#	Overlapp/add method for piecing together the processed windows
		NewSong[i : i + W/2 - 1, j] = NewSong[i : i + W/2 - 1, j] + ProcessedSample[0 : W/2 - 1]
		NewSong[i + W/2 : i + W - 1, j] = ProcessedSample[W/2 : W - 1]

		#	Print proggress
		count = count + 1
		# newprogress = math.floor((100 * count / total) / 10) * 10
		# if not (progress == newprogress) :
		# 	print('\t\tProgress: %d%%' % newprogress)
		
		# progress = newprogress

#	Write clean song
wav.write(output_original_file, Fs, NewSong)
print('Noise reduction and file writing took %.4f seconds' % (time.time() - start_time))
#	Time measure
start_time = time.time()

#	Reconvert to original format
if not input_original_extension in wav_extensions :
	AudioSegment.from_file(output_original_file, input_converted_format).export(output_converted_file, format = output_converted_format)
	print('Conversion to %s took %.4f seconds' % (input_original_format, time.time() - start_time))
	#	Time measure
	start_time = time.time()

File: scripts/matlab/clean_old.m
%   Noise reduction using spectral noise gating.
%   Reads a CSV file containing results of noise analysis, and proceeds
%   to reduce the noise of a song.
%   
%   V1: Noise reduced, but time-smoothing and frequency smoothing not yet
%       implemented.
%
%   Parameters:
%       songpath (String):              Path for input dirty songs
%       noisepath (String):             Path for noise statistics files
%       outputpath (String):            Path for output clean song
%       songname (String):              Name of input dirty song
%       songextension (String):         Song extension
%       noisename (String):             Name of noise to reduce
%       noiseextension (String):        Noise extension
%       W (Integer):                    Window size, in samples
%       MSS (Integer):                  Number of non-overlapped samples
%                                       per window
%       FFTsize (Integer):              Size of FFT
%       ReduceLevel (Integer):          Gain to reduce noise, in dB (+)
%       FreqSmoothingBands (Integer):   Number of freq. smoothing bands
%       TimeSmoothingPerc (Integer):    Percentage allowed for two freqs.
%                                       to differ from consecutive windows
%
%   ======================================================================

%   Time measure;
tic;

%   Reset workspace
close all; clear variables; clc;

%   ======================================================================

%   Independent parameters
Version = '1';
songpath = '..\\Canciones sucias\\';
noisepath = 'profiles\\';
outputpath = '..\\Canciones limpias\\';
songname = 'Raphael - Yo soy aquel';
songextension = '.wav';
noisename = '70s3';
noiseextension = '.csv';

ReduceLevel = 20;
FreqSmoothingBands = 3;
TimeSmoothingPerc = 40;

%   ======================================================================

%   Read files (orignial audio, and noise)
[Song, Fs] = audioread(strcat(songpath, songname, songextension));
NoisePowers = csvread(strcat(noisepath, noisename, noiseextension));
NoisePowers = NoisePowers / 40;

%   Matrix dimensions
s = size(Song);
songchannels = s(2);
songlength = s(1);
n = size(NoisePowers);
FFTsize = n(1);
noisechannels = n(2);

%   Parameters depending on the previous ones
W = FFTsize;
MSS = W / 2;
ReduceLevelUN = 10 ^ ReduceLevel / 10;
OverlappAddBuffer = zeros(MSS, 1);
times = 0;

%   Window
Window = hann(W);

%   ======================================================================

%   Initialize
NewSong = zeros(songlength + W, songchannels);
BufferedSample = zeros(W, songchannels);

%   Print status
fprintf('Processing %d samples\n', songlength);

%   Process song
for j = 1 : songchannels
    
    fprintf('\n\tChannel no. %d\n', j);
    total = floor(songlength / MSS);
    count = 0;
    progress = round(100 * count / total);
    
    for i = 1 : MSS : songlength
        
        %   Print proggress
        count = count + 1;
        newprogress = round(100 * count / total);
        if(progress ~= newprogress)
            fprintf('\t\tProgreso: %d%%\n', newprogress);
        end
        progress = newprogress;
        
        %   Calculate end of song and noise
        songend = min(i + W - 1, songlength);
        
        %   Sample
        SongSample = Song(i : songend, j);
        position = 0;
        
        %   If we reached the end
        if (length(SongSample) < W)
            position = 1;
            %   Fill with zeros until size equals W
            Zeros = zeros(W, 1);
            Zeros(1 : length(SongSample), 1) = SongSample;
            SongSample = Zeros;
        elseif (i == 1)
            position = -1;
        end
        
        %   Windowed sample
        WindowedSample = SongSample .* Window;
        
        %   Compute FFT
        SongTransform = fft(WindowedSample, FFTsize);
        Power = abs(SongTransform) .^ 2;
        
        %   Calculate the power
        for k = 1 : length(Power)
            if Power(k) <= NoisePowers(k, j)
                times = times + 1;
                SongTransform(k) = SongTransform(k) / ReduceLevelUN;
            end
        end
        
        %   Inverse FFT
        ProcessedSample = ifft(SongTransform, FFTsize);
        ProcessedSample = ProcessedSample(1 : W);
        
        %   Save buffered sample
        BufferedSample = ProcessedSample;
        
        %   Overlapp/add method for piecing together the processed windows
        switch position
            case -1
                NewSong(i + W/2 : i + W - 1, j) = zeros(W/2, 1);
                NewSong(i + W : i + 3/2 * W - 1, j) = ...
                    ProcessedSample(1 + W/2 : 1 + W - 1, 1);
            case 1
                NewSong(i + W/2 : i + W - 1, j) = ...
                    NewSong(i + W/2 : i + W - 1, j) + ...
                    ProcessedSample(1 : 1 + W/2 - 1, 1);
                NewSong(i + W : i + 3/2 * W - 1, j) = zeros(W/2, 1);
            otherwise
                NewSong(i + W/2 : i + W - 1, j) = ...
                    NewSong(i + W/2 : i + W - 1, j) + ...
                    ProcessedSample(1 : 1 + W/2 - 1, 1);
                NewSong(i + W : i + 3/2 * W - 1, j) = ...
                    ProcessedSample(1 + W/2 : 1 + W - 1, 1);
        end
    end
end

%   Time smoothing
%   TODO

%   Save result
audiowrite(strcat(outputpath, songname, '_', ...
    noisename, '_', num2str(FFTsize), '_V', Version, songextension), ...
    NewSong, Fs);

%   Time measure
fprintf('\n');
toc;
fprintf('Noise gate applied %d times (%.3f%% of the analyzed bands).\n',...
    times, round(round(times * 1000 / ...
    ((songlength / MSS) * FFTsize))) / 1000);